testing/tests/DevExpress.ui.widgets/button.tests.js

```diff
@@ -542,6 +542,57 @@ QUnit.module('submit behavior', {
         ValidationEngine.registerValidatorInGroup('testGroup', validator);
         this.$element.trigger('dxclick');
     });
+
+    QUnit.test('Form should be submitted only when an async validation rule is passed positively (T887207)', function(assert) {
+        this.clock.restore();
+        let value = 'a';
+        const validValue = 'b';
+        const validator = new Validator($('<div>').appendTo(this.$form), {
+            adapter: sinon.createStubInstance(DefaultAdapter),
+            validationRules: [{
+                type: 'async',
+                validationCallback: function() {
+                    const d = new Deferred();
+                    setTimeout(() => {
+                        d.resolve({
+                            isValid: value === validValue
+                        });
+                    }, 10);
+                    return d.promise();
+                }
+            }]
+        });
+        const done = assert.async();
+        const onSubmit = () => {
+            assert.strictEqual(value, validValue, 'submitted with valid value');
+
+            ValidationEngine.initGroups();
+            this.$form.off('submit', onSubmit);
+            done();
+        };
+        const triggerButtonClick = () => {
+            this.$element.trigger('dxclick');
+        };
+
+        this.$form.on('submit', onSubmit);
+
+        this.$element.dxButton({
+            validationGroup: 'testGroup',
+            onOptionChanged: function(args) {
+                if(args.name === 'disabled') {
+                    if(args.value === false && validator._validationInfo.result.status === 'invalid') {
+                        setTimeout(function() {
+                            value = validValue;
+                            triggerButtonClick();
+                        });
+                    }
+                }
+            }
+        });
+
+        ValidationEngine.registerValidatorInGroup('testGroup', validator);
+        triggerButtonClick();
+    });
```

Comment:
Let's simplify this test.

```suggestion
    QUnit.test('Form should be submitted when an async validation rule is passed positively (T887207)', function(assert) {
        this.clock.restore();

        const done = assert.async();

        const validator = new Validator($('<div>').appendTo(this.$form), {
            adapter: sinon.createStubInstance(DefaultAdapter),
            validationRules: [{
                type: 'async',
                validationCallback: function() {
                    const d = new $.Deferred();
                    setTimeout(() => {
                        d.resolve({ isValid: true });
                    }, 10);
                    return d.promise();
                }
            }]
        });

        ValidationEngine.registerValidatorInGroup('testGroup', validator);

        this.$element.dxButton({ validationGroup: 'testGroup' });

        this.$form.on('submit', () => {
            assert.strictEqual(this.$element.dxButton('option', 'disabled'), false);
            done();
        });

        this.$element.trigger('dxclick');
    });
```

Reply:
The main idea is to check that the form is not submitted when a value is invalid and submitted only when a value is valid. I would not simplify this test to prevent regressions during renovations and so forth.

Reply:
Ok.